[CoSimulationApplication] Adding VTK debugging option for Kratos Mapping

loumalouomega opened the pull request:
**📝 Description**

Adding new mapping class with VTK debuggin, derived from Kratos Mapping.

For example can be sued as follows:

~~~json
"data_transfer_operators" : {
      "mapper" : {
          "type" : "kratos_mapping_with_debug",
          "mapper_settings" : {
              "mapper_type" : "nearest_neighbor"
          },
          "debug_settings" : {
              "auxiliary_debug_counter" : true
          }
      }
  }
~~~

It will generate VTK folder for the pre and post mapping.

**🆕 Changelog**

- [Adding VTK debugging option for Kratos Mapping](https://github.com/KratosMultiphysics/Kratos/commit/039144a465e9d62178dda93a27a75953cbec66a5)
- [Adding auxiliary counter for wrappers without TIME or STEP defined](https://github.com/KratosMultiphysics/Kratos/pull/12511/commits/2045fc93dea24f83b4453939a68e91f4a634ed85)

loumalouomega committed:
[CoSimulationApplication] Adding VTK debugging option for Kratos Mapping

philbucher reviewed (changes requested):
Will do a review on the weekend

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

loumalouomega committed:
Refactor and improvement

loumalouomega committed:
Additional refactor

loumalouomega committed:
Simplify design

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -140,3 +143,208 @@ def __GetRankZeroModelPart():
+            pre_origin_settings["Parameters"]["output_path"].SetString(origin_output_path + "_pre_map")
+
+            # Create a VTK output process object with the provided settings and the current model
+            pre_process_origin = vtk_output_process.Factory(pre_origin_settings, model_origin)
```
I would use the vtk_output directly, then you only need to call `PrintOutput`, and not have all the overhead of the process

loumalouomega replied:
I understand your point, but the problem with that impliementation is that I will need to manually add the settings for the IO instead of automatically deduce the variables and settings from the mapping like I do in the current proposal.

loumalouomega replied:
Ping @philbucher

philbucher reviewed (commented):
I think this is a good addition, but the implementation should be moved to a `CouplingOperation` instead, this is the intended place for such things

Check for example the [CouplingOutput](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CoSimulationApplication/python_scripts/coupling_operations/coupling_output.py) which has a similar purpose

=> then the user can add it to the `coupling_operations` and call it `before_data_transfer_operations` or `after_data_transfer_operations`

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

philbucher commented on applications/CoSimulationApplication/python_scripts/base_classes/co_simulation_data_transfer_operator.py:
```diff
@@ -63,5 +63,6 @@ def _CheckAvailabilityTransferOptions(cls, transfer_options):
             "type"       : "UNSPECIFIED",
-            "echo_level" : 0
+            "echo_level" : 0,
+            "debug_vtk"  : false
```
```suggestion
```

this should not be here

loumalouomega replied:
If not does not work...

loumalouomega committed:
Minor refactor

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

loumalouomega committed:
Adding auxiliary counter for wrappers without TIME or STEP defined

loumalouomega committed:
Refactor KratosMappingDataTransferOperator for improved VTK debugging

The commit refactors the `KratosMappingDataTransferOperator` class in the `CoSimulationApplication` to improve VTK debugging. It adds an auxiliary counter for wrappers without `TIME` or `STEP` defined, and updates the output paths for VTK output processing. The changes ensure that the VTK output files are saved in the correct directory and have appropriate names.

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

loumalouomega committed:
Refactor KratosMappingDataTransferOperator to track the number of mappers and improve VTK debugging

loumalouomega committed:
Set `output_sub_model_parts`

loumalouomega committed:
Extra ,

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

philbucher commented on applications/CoSimulationApplication/python_scripts/base_classes/co_simulation_data_transfer_operator.py:
please move the additions in this file, they should not be in the baseclass

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -140,3 +163,301 @@ def __GetRankZeroModelPart():
+                if self.sub_counter >= 2 * self.number_of_mappers: # 2 because of pre and post mapping
+                    self.sub_counter = 0
+                    self.counter += 1
+                name_prefix = "MAPPING_{}".format(self.counter)
```
```suggestion
                name_prefix = f"MAPPING_{self.counter}"
```

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -140,3 +163,301 @@ def __GetRankZeroModelPart():
+                directory_name = os.path.join(directory_name, path_splitted[i])
+            base_name = path_splitted[-1]
+            if self.auxiliary_debug_counter:
+                directory_name = os.path.join(directory_name, name_prefix)
```
please dont use os.path in modern python, use `pathlib.Path`

loumalouomega replied:
Okay

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -140,3 +163,301 @@ def __GetRankZeroModelPart():
+        vtk_default_output_parameters = KM.Parameters("""{
+            "Parameters" : {
+                "model_part_name"                    : "",
+                "file_format"                        : "ascii",
```
why ascii?

loumalouomega replied:
It is for debugging, we like to take a direct to to the files from time to time

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -37,27 +48,35 @@ def __init__(self, settings, parent_coupled_solver_data_communicator):
-        identifier_tuple         = (identifier_origin, identifier_destination)
-        inverse_identifier_tuple = (identifier_destination, identifier_origin)
+        # Prepare the solver data
+        model_part_origin, model_part_destination, model_part_origin_name, model_part_destination_name, variable_origin, variable_destination, mapper_flags, identifier_origin, identifier_destination, identifier_tuple, inverse_identifier_tuple = self.__PrepareSolverData(from_solver_data, to_solver_data, transfer_options)
```
11 return arguments :sweat_smile:

loumalouomega replied:
![flat,750x,075,f-pad,750x1000,f8f8f8](https://github.com/user-attachments/assets/3ed8ab7d-6d3d-4e6c-935d-fb97045caed1)

loumalouomega replied:
If I move to a new class I can refactor this and make it cleaner perhaps

philbucher reviewed (commented):
I had a look and TBH its very complicated and not very nicely done

How about you derive a class from it, e.g. `kratos_mapping_with_debug_output` to not mess up the baseclass?

loumalouomega committed:
Style

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Merge branch 'master' into CoSim/adding-vtk-debugging-kratos-mapping

loumalouomega committed:
Merge branch 'CoSim/adding-vtk-debugging-kratos-mapping' of https://github.com/KratosMultiphysics/Kratos into CoSim/adding-vtk-debugging-kratos-mapping

loumalouomega committed:
Revert changes in base transfer operator

loumalouomega committed:
Refactor mapping oeprator, remove debug stuff and making some private variables/methods protected

loumalouomega committed:
Update test

loumalouomega committed:
making a standalone class for debug

loumalouomega committed:
Extend documentation

loumalouomega committed:
Refactor mapping operator for debug, remove os dependency and replace with Pathlib

loumalouomega committed:
Refactor and fix test

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -12,10 +12,20 @@
+    'mapper_flags',
+    'identifier_origin', 'identifier_destination',
+    'identifier_tuple', 'inverse_identifier_tuple'
+])
```
please use a `dataclass`  for this => https://docs.python.org/3/library/dataclasses.html

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping_with_debug.py:
```diff
@@ -0,0 +1,379 @@
+        vtk_default_output_parameters = KM.Parameters("""{
+            "Parameters" : {
+                "model_part_name"                    : "",
+                "file_format"                        : "ascii",
```
why ascii?

loumalouomega replied:
Because this is purely for debugging

philbucher replied:
Yes but you want to debug the mapper not the vtk output 

I am not aware of any issues with binary, snd the size difference is considerable, thus I strongly recommend it

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping_with_debug.py:
```diff
@@ -0,0 +1,379 @@
+            self.__debug_vtk_pre = {}
+
+        # Define __debug_vtk_post if not defined
+        if not hasattr(self, "_KratosMappingDataTransferOperatorWithDebug__debug_vtk_post"):
```
I suggest to replace this with a variable that initialized to `None`

loumalouomega replied:
Okay

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping_with_debug.py:
```diff
@@ -0,0 +1,379 @@
+        if self.auxiliary_debug_counter:
+            directory_name /= name_prefix  # Append the name prefix
+        file_path = directory_name / f"post_map_{base_name}"
+        post_destination_settings["Parameters"]["output_path"].SetString(str(file_path))
```
is the `str` needed? I was quite sure that pybind does the conversion automatically

loumalouomega replied:
In give issues in Windows (at least my machine)

philbucher replied:
Fair enough 

Can you post the error please, for reference?

loumalouomega replied:
TypeError: SetString(): incompatible function arguments. The following argument types are supported:
    1. (self: Kratos.Parameters, arg0: str) -> None

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping_with_debug.py:
```diff
@@ -0,0 +1,379 @@
+            model_part_destination = self._GetModelPartFromInterfaceData(to_solver_data)
+
+            if model_part_origin.IsDistributed() or model_part_destination.IsDistributed():
+                mapper_create_fct = python_mapper_factory.CreateMPIMapper
```
why do you override the creation of mappers? 
How about calling the baseclass for this instead?

loumalouomega committed:
Refactor and improve data transfer operator class using dataclass

loumalouomega committed:
Initialize properly

loumalouomega committed:
Reduce code duplication

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/kratos_mapping.py:
```diff
@@ -9,13 +9,28 @@
 from KratosMultiphysics.CoSimulationApplication.utilities import data_communicator_utilities
 from time import time
+from dataclasses import dataclass
+from typing import Union, Tuple
```
ah my archnemesis, typing in python ...

/rant over

loumalouomega replied:
It was your suggestion...

philbucher replied:
I know, its the direction the language goes, smarter ppl than me design this

still, I can complain about it :D

philbucher reviewed (approved):
:+1:
